components/tidb_query_executors/src/index_scan_executor.rs

```diff
@@ -386,123 +432,359 @@ impl IndexScanExecutorImpl {
+    // 3. Depend on the collation if `_bin` or not. Process them differently to get the correct original data.
+    // 4. Write the original data into the column, we need to make sure pop() is called.
+    fn restore_original_data<'a>(
+        &self,
+        restored_values: &[u8],
+        column_iter: impl Iterator<Item = (&'a FieldType, &'a i64, &'a mut LazyBatchColumn)>,
+    ) -> Result<()> {
+        let row = RowSlice::from_bytes(restored_values)?;
+        for (field_type, column_id, column) in column_iter {
+            if !field_type.need_restored_data() {
+                continue;
             }
-            Ok(value[3..].split_at(handle_len))
-        } else {
-            Ok(value.split_at(0))
+            let is_bin_collation = field_type
+                .collation()
+                .map(|col| col == Collation::Utf8Mb4Bin)
+                .unwrap_or(false);
+
+            assert!(!column.is_empty());
+            let mut last_value = column.raw().last().unwrap();
+            let decoded_value = last_value.read_datum()?;
+            if !last_value.is_empty() {
+                return Err(other_err!(
+                    "Unexpected extra bytes: {}",
+                    log_wrappers::Value(last_value)
+                ));
+            }
+            if decoded_value == Datum::Null {
+                continue;
+            }
+            column.mut_raw().pop();
+
+            let original_data = if is_bin_collation {
+                // _bin collation, we need to combine data from key and value to form the original data.
+
+                // Unwrap as checked by `decoded_value.read_datum() == Datum::Null`
+                let truncate_str = decoded_value.as_string()?.unwrap();
+
+                let space_num_data = row
+                    .get(*column_id)?
+                    .ok_or_else(|| other_err!("Unexpected missing column {}", column_id))?;
+                let space_num = decode_v2_u64(space_num_data)?;
+
+                // Form the original data.
+                truncate_str
+                    .iter()
+                    .cloned()
+                    .chain(std::iter::repeat(PADDING_SPACE as _).take(space_num as _))
+                    .collect::<Vec<_>>()
+            } else {
+                let original_data = row
+                    .get(*column_id)?
+                    .ok_or_else(|| other_err!("Unexpected missing column {}", column_id))?;
+                original_data.to_vec()
+            };
+
+            let mut buffer_to_write = column.mut_raw().begin_concat_extend();
+            buffer_to_write.write_v2_as_datum(&original_data, field_type)?;
+        }
+
+        Ok(())
+    }
+
+    // get_index_version is the same as getIndexVersion() in the TiDB repo.
+    fn get_index_version(value: &[u8]) -> Result<i64> {
+        if value.len() <= MAX_OLD_ENCODED_VALUE_LEN {
+            return Ok(0);
+        }
+        let tail_len = value[0] as usize;
+        if tail_len > value.len() {
```

Comment:
```suggestion
        if tail_len > value.len() - 1 {
```

value === [tail_len][ ... ][tail]. It should exclude the tail_len byte.